src/backend/distributed/planner/query_colocation_checker.c

```diff
@@ -168,11 +168,26 @@ AnchorRte(Query *subquery)
 		{
 			Oid relationId = currentRte->relid;
 
-			if (IsCitusTable(relationId) && !HasDistributionKey(relationId))
+			if (!IsCitusTableType(relationId, DISTRIBUTED_TABLE))
 			{
 				/*
-				 * Non-distributed tables should not be the anchor rte since they
-				 * don't have distribution key.
+				 * We're not interested in non distributed relations.
+				 */
+				continue;
+			}
+
+			bool gotDistTableWithShardKey =
+				currentRte->rtekind == RTE_RELATION &&
+				HasDistributionKey(currentRte->relid);
+			bool anchorRteIsDistTableWithoutShardKey =
+				anchorRangeTblEntry &&
+				anchorRangeTblEntry->rtekind == RTE_RELATION &&
+				!HasDistributionKey(anchorRangeTblEntry->relid);
+			if (gotDistTableWithShardKey && anchorRteIsDistTableWithoutShardKey)
+			{
+				/*
+				 * Keep the distributed table that we've found earlier if it
```

Comment:
why do we continue? The currentRTE is a distributed table with dist key, and shouldn't we prefer that becoming the `anchor` rte? Otherwise:
```SQL
explain select * from (SELECT * FROM nullkey_c1_t1) foo JOIN (SELECT * FROM distributed_table) as bar USING (a) JOIN (SELECT * FROM distributed_table) as foo2 USING (a) LIMIT 1;
┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                            QUERY PLAN                                                             │
├───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=0 width=0)                                                                    │
│   ->  Distributed Subplan 44_1                                                                                                    │
│         ->  Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=100000 width=8)                                                   │
│               Task Count: 32                                                                                                      │
│               Tasks Shown: One of 32                                                                                              │
│               ->  Task                                                                                                            │
│                     Node: host=localhost port=9701 dbname=onderkalaci                                                             │
│                     ->  Seq Scan on distributed_table_102014 distributed_table  (cost=0.00..32.60 rows=2260 width=8)              │
│   ->  Distributed Subplan 44_2                                                                                                    │
│         ->  Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=100000 width=8)                                                   │
│               Task Count: 32                                                                                                      │
│               Tasks Shown: One of 32                                                                                              │
│               ->  Task                                                                                                            │
│                     Node: host=localhost port=9701 dbname=onderkalaci                                                             │
│                     ->  Seq Scan on distributed_table_102014 distributed_table  (cost=0.00..32.60 rows=2260 width=8)              │
│   Task Count: 1                                                                                                                   │
│   Tasks Shown: All                                                                                                                │
│   ->  Task                                                                                                                        │
│         Node: host=localhost port=9702 dbname=onderkalaci                                                                         │
│         ->  Limit  (cost=0.01..5.11 rows=1 width=16)                                                                              │
│               ->  Nested Loop  (cost=0.01..16970.17 rows=3325 width=16)                                                           │
│                     Join Filter: (nullkey_c1_t1.a = intermediate_result_1.a)                                                      │
│                     ->  Nested Loop  (cost=0.00..2007.66 rows=665 width=16)                                                       │
│                           Join Filter: (nullkey_c1_t1.a = intermediate_result.a)                                                  │
│                           ->  Function Scan on read_intermediate_result intermediate_result  (cost=0.00..10.00 rows=1000 width=8) │
│                           ->  Materialize  (cost=0.00..3.00 rows=133 width=8)                                                     │
│                                 ->  Seq Scan on nullkey_c1_t1_102008 nullkey_c1_t1  (cost=0.00..2.33 rows=133 width=8)            │
│                     ->  Function Scan on read_intermediate_result intermediate_result_1  (cost=0.00..10.00 rows=1000 width=8)     │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
(28 rows)
```

Overall, I think we really should simplify the logic here. The AnchorRTE selection is NOT optimal anyway, and we are not gaining much here. So, can you just get back to the original code:

```SQL
		else if (currentRte->rtekind == RTE_RELATION)
		{
			Oid relationId = currentRte->relid;

			if (!IsCitusTableType(relationId, DISTRIBUTED_TABLE)
			{
				/*
				 * We're not interested in non distributed relations.
				 */
				continue;
			}

			anchorRangeTblEntry = currentRte;
			break;
		}
```

At somepoint we might optimize this code, for now, lets just keep it simple

Reply:
For testing, play with queries in the following form and make sure anything non-colocated gets recursively planned:
```SQL
explain select * from (SELECT * FROM nullkey_c1_t1) foo JOIN (SELECT * FROM distributed_table) as bar USING (a) JOIN (SELECT * FROM distributed_table) as foo2 USING (a) LIMIT 1;

explain select * from nullkey_c1_t1 foo JOIN  distributed_table as bar USING (a) JOIN (SELECT * FROM distributed_table) as foo2 USING (a) LIMIT 1;

```

where change the orders of foo/bar/foo2 and sometimes use them directly as tables sometimes inside subqueries. 

Or, make subqueries non-colocated by using non-dist keys joining -- see `b as a` in the target list:

```SQL
explain select * from (SELECT * FROM nullkey_c1_t1) foo JOIN distributed_table as bar USING (a) JOIN (SELECT b as a FROM distributed_table) as foo2 USING (a) LIMIT 1;
```

Reply:
Some other example tests cases:
https://github.com/citusdata/citus/blob/main/src/test/regress/sql/having_subquery.sql
https://github.com/citusdata/citus/blob/main/src/test/regress/sql/subquery_and_cte.sql
https://github.com/citusdata/citus/commit/f2538a456f075f15e475e87260c99062992dd69d#diff-db780b730b14664acd9e7ae3c0965f35011274cc25c6114f812467198b0ca921R5

Reply:
* try a few different arbitrary config tests, such as single node?
* update delete using complex subqueries

Reply:
Note to myself: only remaining thing to check is the last one:
> https://github.com/citusdata/citus/commit/f2538a456f075f15e475e87260c99062992dd69d#diff-db780b730b14664acd9e7ae3c0965f35011274cc25c6114f812467198b0ca921R5

Reply:
done